029-across-step/proposal.md

```diff
@@ -0,0 +1,190 @@
+# `across` step
+
+This proposal introduces a mechanism by which a build plan can be executed
+multiple times with different `((vars))`, in service of build matxies, pipeline
+matrixes, and - building on [`((var))` sources][var-sources-rfc] - dynamic
+variations of both.
+
+
+## Motivation
+
+* Support dynamic multi-branch workflows:
+  [concourse/concourse#1172][multi-branch-issue]
+* Support build matrixes and pipeline matrixes.
+* Replace a common, and very flawed, usage pattern of `version: every`, which
+  has proven to be very complicated to support.
+
+
+## Proposal
+
+The `across` step is given a set of values, a name for a `((var))`, and a
+sub-plan to execute with each value assigned as the named var.
+
+The set of values can be static, as below:
+
+```yaml
+across: [1.12, 1.13]
+as: go-version
+do:
+- task: unit
+  vars: {go_version: ((go_version))}
+```
+
+In keeping with `((var))` semantics, values may also be complex types.
+
+Their fields can be accessed off the var or passed verbatim, as in the
+following examples:
+
+```yaml
+across:
+- from: 3.0
+  to: 5.0
+- from: 4.0
+  to: 5.0
+as: upgrade_path
+do:
+- set_pipeline: upgrade-test
+  instance_vars:
+    upgrade_from: ((upgrade_path.from))
+    upgrade_to: ((upgrade_path.to))
+```
+
+```yaml
+across:
+- upgrade_from: 3.0
+  upgrade_to: 5.0
+- upgrade_from: 4.0
+  upgrade_to: 5.0
+as: upgrade_path
+do:
+- set_pipeline: upgrade-test
+  instance_vars: ((upgrade_path))
+```
+
+### Dynamically iterating over `((vars))` from `var_sources`
+
+Instead of static values, a [`((var))` source][var-sources-rfc] may be
+referenced by name in order to execute the plan across dynamic `((vars))` at
+runtime.
+
+This gives full control of parameterization to the user. For example, a
+`github-prs` prototype could be implemented which returns a `((var))` for each
+pull request, or the `git` prototype could be used as a var source to provide
+each branch as a `((var))`.
+
+```yaml
+var_sources:
+- name: booklit-prs
+  type: github-prs
+  config:
+    repository: vito/booklit
+    access_token: # ...
+
+plan:
+- across: booklit-prs
+  as: pr
+  do:
+  - set_pipeline: pr
+    instance_vars: {pr_number: ((pr.number))}
+```
+
+### Triggering on changes
+
+With `trigger: true` configured, the build plan will run on any change to the
+set of `((vars))` - i.e. when a var is added or removed.
+
+
+## Open Questions
+
+* It seems like triggering could work in both the static and dynamic cases, but
```

Comment:
Assuming I understand the difference between static vs dynamic, I could see static triggers as empty payloads having value for things like smoke tests, where you want to continuously test your system, so you want a static trigger where the value of the trigger is irrelevant (i.e. time resource), you just need the trigger to have fired. So let's say that I have two identical (for the sake of argument) environments, beta and gamma, and I want to run scheduled smoke tests, I could see it being something like this:

```yaml
resources:
- name: 5m
  type: time
  source: {interval: 5m}

plan:
- across: [beta, gamma]
  as: 5m
  do:
  - task: smoke-tests
    file: smoke-tests-task.yml
```

Again, assuming I understand the semantics of dynamic vs static, that's where my thought process lead. In this case the value of any of the inputs is diminished because the task config is not needed due to some default input in the task, the value of `across` in this case is that the build matrix was executed statelessly. Does that makes sense?

Reply:
@mxplusb Sorry, didn't see this by the time I got back to this RFC!

I think your example above would be written like this:

```yaml
resources:
- name: 5m
  type: time
  source: {interval: 5m}

plan:
- get_var: 5m
  trigger: true
- across:
  - var: env
    values: [beta, gamma]
  do:
  - task: smoke-tests
    file: smoke-tests-task.yml
```

This question was more about whether it makes sense to configure `trigger: true` on the `var: env` part. The reason it's an interesting case is that the values will only ever change when the config changes, so it's a little weird to 'trigger' on it, but I decided we should support it since it still makes sense from the perspective of 'I changed the config, and I want a change to this config to auto-trigger a new build'.

re: your other comment (`max_in_flight`), that has been added to the proposal. :)